### PR TITLE
integration: added HaystackAction

### PR DESCRIPTION
**Pull request by zilto**
Add `HaystackAction` which allows to define a Burr `Action` from a Haystack `Component`.

## Changes
- `HaystackAction` class
- `haystack_pipeline_to_burr_graph()` function
- added a tutorial notebook detailing some motivation and illustrating code with and without the integration

## How I tested this
- no test yet

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

----

> [!IMPORTANT]
> Add `HaystackAction` and conversion function to integrate Haystack components and pipelines with Burr.
> 
>   - **Integration**:
>     - Add `HaystackAction` class in `haystack.py` to convert Haystack `Component` to Burr `Action`.
>     - Add `haystack_pipeline_to_burr_graph()` function in `haystack.py` to convert Haystack `Pipeline` to Burr `Graph`.
>   - **Documentation**:
>     - Add `haystack.rst` for integration reference.
>     - Add tutorial in `examples/haystack-integration/README.md`.
>   - **Examples**:
>     - Add `application.py` demonstrating integration usage.
>   - **Tests**:
>     - Add `test_burr_haystack.py` to test `HaystackAction` and pipeline conversion.
>   - **Dependencies**:
>     - Add `haystack` to `pyproject.toml` under optional dependencies.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=DAGWorks-Inc%2Fburr&utm_source=github&utm_medium=referral)<sup> for c2927b23fb92696dfd94da1d2ecf6d4225d3b003. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: burr/integrations/haystack.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    def __init__(
+        self,
+        component: Component,
+        reads: Union[list[str], dict[str, str]],
+        writes: Union[list[str], dict[str, str]],
+        name: Optional[str] = None,
+        bound_params: Optional[dict] = None,
+    ):
+        """
```
Comment (elijahbenizzy) [inline burr/integrations/haystack.py:23]: More docstrings

File: burr/integrations/haystack.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        """Update the state using the results of `Component.run()`."""
+        state_update = {}
+        for socket_name, value in result.items():
+            state_field_name = self._socket_mapping.get(socket_name, socket_name)
+            if state_field_name in self.writes:
+                state_update[state_field_name] = value
+
+        return state.update(**state_update)
+
+    def bind(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:104]: Otherwise we can make this a more specific name to not conflict

[the remark this answers is not in the document]
Comment (zilto) [reply burr/integrations/haystack.py:104]: If I understand, the pattern is to only allow `bound_params` at init?

This would be ok:
```python
haystack_retrieve_documents = HaystackAction(
    component=InMemoryEmbeddingRetriever(InMemoryDocumentStore()),
    name="retrieve_documents",
    reads=["query_embedding"],
    writes=["documents"],
    bound_params={"foo": "bar"},
)
```
This would not be ok?
```python
haystack_retrieve_documents = HaystackAction(
    component=InMemoryEmbeddingRetriever(InMemoryDocumentStore()),
    name="retrieve_documents",
    reads=["query_embedding"],
    writes=["documents"],
)
haystack_retrieve_documents.bind(foo="bar")
```

I should simply remove the `.bind()` method?

[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:104]: I think so, or call it something else? Alternatively, we can make bind work at the action level. Currently it's only for functions, and class-based actions don't have it.

File: burr/integrations/haystack.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            socket.name for socket in data.get("output_sockets", {}).values() if socket.receivers
+        ]
+        for name, data in pipeline.graph.nodes(data=True)
+    }
+
+
+def haystack_pipeline_to_burr_graph(pipeline: Pipeline) -> Graph:
+    """Convert a Haystack `Pipeline` to a Burr `Graph`.
+
+    From the Haystack `Pipeline`, we can easily retrieve transitions.
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:170]: Nice, I thought that it hd parallelizing capabilities, but it looks like it doesn't?

[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:170]: Ahh, yeah, this won't work: 
image

[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:170]: But after parallelism it will, and we might be able to it to work now

File: burr/integrations/haystack.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    @property
+    def reads(self) -> list[str]:
+        return self._reads
+
+    @property
+    def writes(self) -> list[str]:
+        return self._writes
+
+    @property
+    def inputs(self) -> tuple[dict[str, str], dict[str, str]]:
```
[the remark this answers is not in the document]
Comment (zilto) [reply burr/integrations/haystack.py:52]: Regarding `.inputs()`, it is currently a property, but it's return value may change if we allow a `.bind()` method. What was previously a required/optional input is now bound.

It seems that this logic belongs to `.inputs()` and wouldn't be of much use elsewhere

[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:52]: Yeah, `.bound()` should return a copy of the object so the `.inputs()` are frozen IMO. That said, this should take in `bound_inputs` in the constructor, and we don't need a method? You don't really want properties dynamically computed like this, it's hard to debug and a bit iffy IMO.

File: burr/integrations/haystack.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py:170]: We should add a check to see if this is in there, or label this as experimental. At least document the caveats.

File: burr/integrations/haystack.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    ):
+        """
+        Notes
+        - need to figure out how to use bind
+        - you can use `action.bind()` to set values of `Component.run()`.
+        """
+        self._component = component
+        self._name = name
+        self._reads = list(reads.keys()) if isinstance(reads, dict) else reads
+        self._writes = list(writes.values()) if isinstance(writes, dict) else writes
```
[the remark this answers is not in the document]
Comment (zilto) [reply burr/integrations/haystack.py]: I changed some mappings to ensure `reads={socket: state_field}` and `writes={state_field: socket}`.

This seems to make the most sense because:
- on `reads`, each key should be unique and map to a unique kwarg of `Component.run()`. The same state value could be passed to different kwargs
- on `writes`, each key should be a unique state field. Even though `Component.run()` returns a dictionary where keys are unique, we want to prevent calling `.update()` on the same field several times.

In other words, this would be invalid:

```python
class CustomComponent:
  def run() -> dict:
     return {"bar": 1, "baz": 2}

# mapping both `bar` and `baz` to `State["foo"]` is invalid
HaystackAction(
   component=CustomComponent()
   writes={"bar": "foo", "baz": "foo"}
)

# reverting the mapping ensures that two outputs can't be mapped to `foo`
HaystackAction(
   component=CustomComponent()
   writes={"foo1": "bar", "foo2": "baz"}
)
```

[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py]: Yes, I think that makes sense, just make sure it's thoroughly documented

File: burr/integrations/haystack.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply burr/integrations/haystack.py]: So yeah, the pattern should be a copy-operation here, or remove this and just do the inputs in the parameters.

**Review by elijahbenizzy — COMMENTED** [review]
One quick point about `bind()` doing a mutation, otherwise looks good

File: burr/integrations/haystack.py [hunk 6]
```diff
@@ -68,47 +142,68 @@ def inputs(self) -> tuple[dict[str, str], dict[str, str]]:
...
 
+        for state_field_name, output_socket_name in self._output_socket_mapping.items():
+            if state_field_name in self.writes:
+                try:
+                    state_update[state_field_name] = result[output_socket_name]
+                except KeyError as e:
+                    raise ValueError(
+                        f"Socket `{output_socket_name}` missing from output of `Component.run()`"
+                    ) from e
+        print(state_update)
```
[the remark this answers is not in the document]
Comment (zilto) [reply burr/integrations/haystack.py]: Something odd happened with the review because the current version doesn't have the `print()` and I removed `.bind()` as we've discussed. Also, docstrings are much more complete

the comments appear as `Outdated`

**Review by elijahbenizzy — APPROVED** [review]
